Enums

ZThatWhiteGuyZ opened the pull request:
Work in progress. More work will be done over the course of the next few days.

ZThatWhiteGuyZ committed:
add enum-tutorial.md

ZThatWhiteGuyZ committed:
update enum-tutorial.md

Commit:
Update enum-tutorial.md

Committing due to weather.

Commit:
Update enum-tutorial.md

Made enums folder and moved tutorial in. Wrote the first part of the introduction. More research required.

Commit:
Update enum-tutorial.md

Added another section. Still a work in progress, but open to suggestions. Pretty hard topic to write on.

Commit:
Update enum-tutorial.md

Added section on errors and duplicate values.

Commit:
Update enum-tutorial.md

Accidentally learned a lot more about enums, so added quite a bit. Variable declaration, typedef, and some rearranging of information.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
````diff
@@ -0,0 +1,168 @@
+We can even define multiple variables of type `exp_value` like this:
+
+```c
+enum exp_value {wheat = 2, onion, carrot, corn = 7, potatoes, strawberries = 11} plants;
+enum exp_value crops, dirt_growers;
````
To keep continuity with the previous example, I may change this bit to have all three instances of `enum exp_value` declared on the same line.

```suggestion
enum exp_value {wheat = 2, onion, carrot, corn = 7, potatoes, strawberries = 11};
enum exp_value crops, plants, dirt_growers;
```

ZThatWhiteGuyZ replied:
This is done for the sake of the example, specifically because we can do either or both.

ZThatWhiteGuyZ replied:
This has been corrected and explained in the tutorial. I also added a "bad practices" section like I mentioned in a previous comment.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
```diff
@@ -0,0 +1,168 @@
+## Using Enums in Code
+
+When using enums, we must define a variable type. This can be done in one of two ways. I will reuse the `exp_value` example.
+
```
Here it would possibly be good to add a generic version of the syntax first, similar to line 11.

i.e.:
``` c
enum <enum_name> {[enumerator1, enumerator2, ...]};
enum <enum_name> <variable_name>;

// or

enum <enum_name> {[enumerator1, enumerator2, ...]} <variable_name>;
```

ZThatWhiteGuyZ replied:
That would make it more readable, so I'll add that in.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
```diff
@@ -0,0 +1,168 @@
+{
+	crops = onion;
+	plants = corn;
+	dirt_growers = potatoes;
```
I may suggest moving the variable declarations to `main()` since generally global variables are considered not-as-good practice.

```suggestion
enum exp_value {wheat = 2, onion, carrot, corn = 7, potatoes, strawberries = 11};

int main(void)
{
	enum exp_value crops = onion;
	enum exp_value plants = corn;
	enum exp_value dirt_growers = potatoes;
```

ZThatWhiteGuyZ replied:
Alternatively, can I add a disclaimer to not declare global variable because it's generally bad practice?

ZThatWhiteGuyZ replied:
In person conversation has shed new light on why this is a bad idea. People like to hard scroll and copy/paste from resources. I will add a "bad practice with enums" section to do this bad stuff.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
````diff
@@ -0,0 +1,168 @@
+
+``` c
+int x = 8
+crops Mystery_crop = x;
````
I feel like this might break some continuity since `crops` was previously a variable name, but now `crops` is being used as a `typedef`'ed enum name. Perhaps `typedef` to `exp_value` instead, in the similar way that structs are sometimes `typedef`'ed to the same name.

Also instead of saying "get a bit more tricky" I would just say that you aren't explicitly using the `enum` keyword anymore.

ZThatWhiteGuyZ replied:
Both of these have been done. Will resolve conversation.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
```diff
@@ -0,0 +1,168 @@
+int main(void)
+{
+	int x = 9;
+	crops Mystery_crop = x;
```
Same note about continuity with the `crops` name. I may change the name `crops` since it was used as a variable name previously.

ZThatWhiteGuyZ replied:
Resolved a long time ago.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
````diff
@@ -0,0 +1,168 @@
+``` c
+#include <stdio.h>
+
+enum days_of_the_week {sun = 1, mon, tues, wed, thur, fri, sat} day;
+enum celsetial_bodies {sun, moon, jupiter} body;
+
+int main(void)
+{
+	day = mon;
+	printf("%d\n", mon);
+	body = sun;
````
Same note about creating global variables.

ZThatWhiteGuyZ replied:
Corrected this, and the enumerator syntax.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
```diff
@@ -0,0 +1,168 @@
+
+The compiler will warn you that it is expecting a semicolon after `crops` if you try to define an enumerator variable this way without the use of `typedef`.
+
+### Using `typedef` With Enums
```
I may move this heading to right before "But what if we want to get a bit more tricky"

ZThatWhiteGuyZ replied:
That's actually what I had originally. Happy to go back to it.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
```diff
@@ -0,0 +1,168 @@
+The compiler will warn you that it is expecting a semicolon after `crops` if you try to define an enumerator variable this way without the use of `typedef`.
+
+### Using `typedef` With Enums
+
+We can use the `typedef` keyword to make our enumerated list a bit more versatile. For example, we will be able to create an enum variable, rather than being restricted to the constants in the enumerated list. Let's look at this example.
+
```
Like the other suggestion, I may first write the generic version to show what the syntax is.

i.e.:
``` c
typedef <enum_definition> <enum_alias>;

// or

typedef enum <enum_name> {[enumerator1, enumerator2, ...]} <enum_alias>;
```

ZThatWhiteGuyZ replied:
Did that, too. Thank you.

jerrettl reviewed (commented):
A lot of really good information here. I learned a good bit about enums from reading through this. I have a few suggestions on semantics for you to make it super clear what's happening on syntax. I hope this isn't too nitpick-y, I am just a big syntax person. Further details ahead:

Commit:
Update enum-tutorial.md

Small update with naming convention fixes. Not complete.

Commit:
Update enum-tutorial.md

A handful of changes. Fixed the rest of the naming convention issues, added a note about splitting up enum variable types (gross), and adjusted the typedef name in the typedef section.

Did not mention the thing about global variable best practice. This can be added if needed, however.

Commit:
Update enum-tutorial.md

Hopefully the final battle.

Commit:
Enum Practice Problem and Answer

Made a single Enum practice problem that I've had in my head since working on the tutorial. Uses typedef.

Commit:
Update enum-practice.md

Added note about stdlib.h for the program to run without implicit declaration.

Commit:
Update enum-tutorial.md

Added the `typedef` syntax for enums. I gotta stop working on this it just keeps getting more wild.

Commit:
Tutorial wording catch

Reread through my enums tutorial and found an funky sentence. Small edit.

Commit:
Update enum-tutorial.md

Lots of updates here. Some syntax corrections, a lot of best practices fixes, language updates,  and a new section on what not to do. Hungry for feedback.

ZThatWhiteGuyZ commented:
Tutorial is complete. Please provide feedback.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
````diff
@@ -0,0 +1,246 @@
+enum days_of_the_week {Sunday = 1, Monday, Tuesday, Wednesday, Thursday, Friday, Saturday};
+```
+
+This assigns `Sunday` the value of 1, and the following value will be 1 greater. In this example, the compiler assigns value to the constants based on the defined value in the form previous + 1. You can define any number of values in an enumerated list and this rule will be followed. For example, we can define the value for each enumerated constant if we choose like this:
````
```suggestion
This assigns `Sunday` the value of 1, and the following value will be 1 greater. In this example, the compiler assigns value to the constants based on the defined value in the form "previous + 1." You can define any number of values in an enumerated list and this rule will be followed. For example, we can define the value for each enumerated constant if we choose like this:
```

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
````diff
@@ -0,0 +1,246 @@
+
+We can use the `typedef` keyword to make our enumerated list a bit more versatile. The basic syntax would look something like this:
+
+``` c
+// 1)
+typedef enum {Const1, Const2, Const3,... ConstN} variable_name;
+
+// or
+
+// 2)
+typedef enum name {Const1, Const2, Const3,... ConstN} variable_name;
+
+// or
+
+// 3)
+enum name {Const1, Const2, Const3,... ConstN};
+typedef enum name variable_name
+```
````
Small thought. Since this already has the "#)" numbering, I would either take out the "// or"s or take out the numbering, since they are both saying the same thing.

```suggestion
``` c
// Variation 1
typedef enum {Const1, Const2, Const3,... ConstN} variable_name;

// Variation 2
typedef enum name {Const1, Const2, Const3,... ConstN} variable_name;

// Variation 3
enum name {Const1, Const2, Const3,... ConstN};
typedef enum name variable_name
```
```

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
```diff
@@ -0,0 +1,246 @@
+int main(void)
+{
+	int x = 9;
+	typedef enum exp_value crop_experience;
```
Maybe we already talked about this but I believe `typedef` statements should be written next to where the enum was defined, since it is globally-defined.

jerrettl commented on src/intro-to-c/enums/enum-tutorial.md:
```diff
@@ -0,0 +1,246 @@
+
+## Bad-Practice Uses of Enums
+
+In general, you shouldn't have global variables. However, enums provide a lot of versatility and wiggle room for things that will work. You may notice this being very similar to the [dirt grower example](#dirt-grower). The only difference is that `crops`, `plants`, and `dirt_growers` are now globally defined.
```
This link doesn't work 😩

jerrettl commented on src/intro-to-c/enums/enum-practice-answers.md:
````diff
@@ -0,0 +1,81 @@
+author: Matthew Gibbons
+---
+
+@. You are a half decent farmer in a video game that received a handful of seeds in the mail. With them came a note that informed you they all take the same amount of time to grow. You decided to plant them all and nurture them to maturity. The day has come to harvest, and you need to keep track of the crops you recognize. Write a program that will keep a count of each crop you have harvested, in ...
+
+``` c
+// Random seed
+srand(24164132);
+// Random number generated
+1 + rand() % 13;
+// Crop Values
+3, 5, 6, 9, 11, 12;
+```
````
If it will use elements of randomness, I would explicitly write in the question that you will also need to generate random crops, and perhaps which crops are the ones that are being farmed.

jerrettl reviewed (commented):
Got a chance to take a look. Thank you for all the work on this particularly crunchy section. All I would say is that (as we have previously discussed), the enumeration naming conventions give me the willies. I believe some of the conventions are inconsistent within the same document, so it may want to be double-checked.

I looked up the enumeration section from the GNU C manual (probably the only thing close to an "official" style guide), and they use snake_case.
https://www.gnu.org/software/gnu-c-manual/gnu-c-manual.html#Enumerations

If some style guide for C can be shown that makes sense and is consistent, I will accept it, but since these pages will essentially act as a "golden standard" for how enums should be written in C, I don't know if I can accept any hint at inconsistency in good conscience. I'm just picky like that 🙃

ALSO don't forget to add these pages to the table of contents.

Thank you for all the work again and dealing with my continuous argument on how enums in C are one of humanity's worst creations.

Other than that, just some small spots that stuck out to me as I was looking through:
